### ISSUE TITLE
Recurrent model (GRU) doesn't learn anything

### ISSUE THREAD

**Issue by optimass** [blei-lab/edward#694]
Hi,

I've implemented a pretty basic RNN (GRU) in Edward that encodes a sequence into latent variables and then decodes it for sentiment analysis (i'm using the IMDB dataset).
The model doesn't seem to learn anything. It seems that the gradient doesn't flow to the weights of the recurrent model (I've taken the look at the updates, and they are always null). 
I'm wondering if it is related to [this issue](https://github.com/blei-lab/edward/issues/525) 
Anyways, here is the code:

```
with tf.variable_scope('model', reuse=True):

    ##### params ######
    W = tf.Variable(tf.random_normal([3, H, H]), name='W', dtype=tf.float32)
    U = tf.Variable(tf.random_normal([3, D, H]), name='U', dtype=tf.float32)
    b = tf.Variable(tf.random_normal([H]),       name='b', dtype=tf.float32)
    ####################
    
    y_ph = tf.placeholder(tf.int32, [batch_size, 1], name='y_ph')

    x = tf.placeholder(tf.int32, [batch_size, nb_steps ], name='x')     

    def gru_cell(hprev, xt):
        #  update gate
        z = tf.sigmoid(tf.matmul(xt,U[0]) + tf.matmul(hprev,W[0]) )
        #  reset gate
        r = tf.sigmoid(tf.matmul(xt,U[1]) + tf.matmul(hprev,W[1]) )
        #  intermediate
        h = tf.tanh(tf.matmul(xt,U[2])    + tf.matmul( (r*hprev),W[2]) )
        # new state
        return (1-z)*h + (z*hprev)
      
    # embed -->
    x_ = Embedding(V, D, name='Embedding')(x)
    
    # initial state of RNN
    h = tf.zeros(shape=(batch_size, H)) # initial state

    for t in range(nb_steps-1):
        h = gru_cell(h, x_[:,t,:])

    ############# Varitational Inference #########################    
    qz = Normal(loc=Dense(E, name='Dense_qz_mean')(h),
               scale=Dense(E, name='Dense_qz_scale',activation='softplus')(h))
        
    z = Normal(loc=tf.zeros([M, E]), scale=tf.ones([M, E]))    
    #######################################################   
    
    y = Dense(1, activation='sigmoid', name='Dense_y')(z.value())
    y = Bernoulli(y)

inference = ed.KLqp({z: qz}, data={y: y_ph})

### Run -->
optimizer = tf.train.RMSPropOptimizer(0.01, epsilon=1.0)
inference.initialize(optimizer=optimizer) # always redefine inference before
sess = ed.get_session()
tf.global_variables_initializer().run()

n_epoch = 100
n_iter_per_epoch = 100

for epoch in range(n_epoch):
  avg_loss = 0.0

  pbar = Progbar(n_iter_per_epoch)
  for t in range(1, n_iter_per_epoch + 1):
    pbar.update(t)   
    batch = np.random.randint(0, len(X_train)-1, batch_size)
    info_dict = inference.update({x: X_train[batch], y_ph: y_train[batch]})
    avg_loss  += info_dict['loss']
    
  # Print a lower bound to the average marginal likelihood for an
  # image.
  avg_loss = avg_loss / n_iter_per_epoch
  avg_loss = avg_loss / batch_size
  print("log p(x) >= {:0.3f}".format(avg_loss))

**Comment by dustinvtran** [comment]
Adding the following to the top of your script, the loss seems to go down.
```python
from edward.models import *
from edward.util import Progbar
import edward as ed
import numpy as np
import tensorflow as tf
from keras.layers import Embedding, Dense

H = 5
D = 2
V = 10
E = 2
batch_size = 10
M = 10
nb_steps = 5

X_train = np.zeros([10000, nb_steps ], dtype=np.int32)
y_train = np.zeros([10000, 1 ], dtype=np.int32)
```
My guess is that more careful tuning of hyperparameters/inits are needed. As a baseline which is simple and stabler, have you tried just MLE/MAP? You would replace `qz` with
```python
qz = PointMass(params=Dense(E, name='Dense_qz_mean')(h))
```
This loss function is just the expected log-joint, which reduces to reconstruction cost plus a penalty:
```
max_{q, theta'} E_{q(z)} [ log p(x, z; theta') ] = 
  max_{z', theta'} log p(x | z'; theta') + log p(z').
```

> I'm wondering if it is related to this issue

That issue is for TensorFlow while loops; unrolling the graph as you do here works fine.

**Comment by optimass** [comment]
Hi Dustin,

Thanks for your help!
Actually adding: 
`scale={y: len(X_train) / batch_size}`
to the initialization did the trick.

You can close/delete the issue

**Comment by dustinvtran** [comment]
Great!

**Comment by wset2** [comment]
Hi

In this example, how do you test that the GRU is working as desired?
